crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -454,12 +510,110 @@ impl Store {
         self.save_changes(changes).await
     }
 
+    fn collect_device_updates(
+        &self,
+        own_identity: Option<ReadOnlyOwnUserIdentity>,
+        identities: &IdentityUpdates,
+        changes: &Changes,
+    ) -> DeviceUpdates {
+        let mut new: BTreeMap<_, BTreeMap<_, _>> = BTreeMap::new();
+        let mut changed: BTreeMap<_, BTreeMap<_, _>> = BTreeMap::new();
+
+        for device in &changes.devices.new {
+            let device = Device {
+                inner: device.to_owned(),
+                verification_machine: self.inner.verification_machine.to_owned(),
+                own_identity: own_identity.to_owned(),
+                device_owner_identity: identities.get_inner(device.user_id()),
+            };
+
+            new.entry(device.user_id().to_owned())
+                .or_default()
+                .insert(device.device_id().to_owned(), device);
+        }
+
+        for device in &changes.devices.changed {
+            let device = Device {
+                inner: device.to_owned(),
+                verification_machine: self.inner.verification_machine.to_owned(),
+                own_identity: own_identity.to_owned(),
+                device_owner_identity: identities.get_inner(device.user_id()),
+            };
+
+            changed
+                .entry(device.user_id().to_owned())
+                .or_default()
+                .insert(device.device_id().to_owned(), device.to_owned());
+        }
+
+        DeviceUpdates { new, changed }
+    }
+
+    async fn collect_identity_updates(
+        &self,
+        changes: &Changes,
+    ) -> Result<(Option<ReadOnlyOwnUserIdentity>, IdentityUpdates)> {
+        let mut new = BTreeMap::new();
+        let mut changed = BTreeMap::new();
+
+        for identity in &changes.identities.new {
+            new.insert(identity.user_id().to_owned(), identity.to_owned());
+        }
+
+        for identity in &changes.identities.changed {
+            changed.insert(identity.user_id().to_owned(), identity.to_owned());
+        }
+
+        let own_identity = if let Some(identity) = new.get(self.user_id()) {
+            identity.own().cloned()
+        } else if let Some(identity) = changed.get(self.user_id()) {
+            identity.own().cloned()
+        } else {
+            self.get_user_identity(self.user_id()).await?.and_then(|u| u.own().cloned())
+        };
+
+        let new = new
+            .into_iter()
+            .map(|(user_id, identity)| {
+                (
+                    user_id,
+                    UserIdentities::new(
+                        identity,
+                        self.inner.verification_machine.to_owned(),
+                        own_identity.to_owned(),
+                    ),
+                )
+            })
```

Comment:
Because we first need to find our own up-to-date user identity to construct `UserIdentities`, which we might take from the new maps, in case our own identity got updated as well, or might need to fetch from the store otherwise.

Reply:
Ohh, I see. The closure captures `own_identity`. But the `let own_identity = ...` could be moved to before the two loops, no?

Reply:
Well we have the identities in a `Vec`, the loops sort them into a map where it then becomes easy to find our own identity.

We would need to search the `Vec` to move the `let own_identity = ...` before the loops.

Reply:
Ugh sorry, I'm finally seeing it, the `let own_identity` uses the result of the two for loops right after the `=` of the two `if let` conditions. Sorry for the noise.

Reply:
It is a bit confusing :sweat_smile:. But due to #2515 this needs to change and be simplified anyways.